### PR TITLE
Markdown writer

### PR DESCRIPTION
**Pull request by erri120**
Part of #327.

Allows value formatters to write either plain text or markdown depending on the state.

This PR doesn't include a markdown renderer.

### CODE REVIEW

File: src/Abstractions/NexusMods.Abstractions.Games.Diagnostics/Writer/DiagnosticWriterMode.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+using JetBrains.Annotations;
+
+namespace NexusMods.Abstractions.Diagnostics;
+
+/// <summary>
+/// Represents the valid modes of a <see cref="IDiagnosticWriter"/>.
+/// </summary>
+[PublicAPI]
+public enum DiagnosticWriterMode
```
Comment (Sewer56) [inline src/Abstractions/NexusMods.Abstractions.Games.Diagnostics/Writer/DiagnosticWriterMode.cs:9]: Should be marked as `byte`.

File: src/Abstractions/NexusMods.Abstractions.Games.Diagnostics/Writer/DiagnosticWriterState.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    /// <summary>
+    /// Gets the mode of the writer.
+    /// </summary>
+    public readonly DiagnosticWriterMode Mode;
+
+    /// <summary>
+    /// Constructor.
+    /// </summary>
+    public DiagnosticWriterState(DiagnosticWriterMode mode)
```
Comment (Sewer56) [inline src/Abstractions/NexusMods.Abstractions.Games.Diagnostics/Writer/DiagnosticWriterState.cs:25]: An optional builder capacity might be handy here.

**Review by Sewer56 — CHANGES_REQUESTED** [review]
This is fine. Just extremely minor nitpicks.

Quick Question:
- Is the `Mode` intended to be dynamically mutated?

Usually for code like this, I see a design where you have a writer for a specific 'mode' and then you invoke methods for that. This isn't a request to change the code, just curious more than anything.

File: src/Abstractions/NexusMods.Abstractions.Games.Diagnostics/Writer/DiagnosticWriterMode.cs (hunk 1 above)
Comment (erri120) [reply src/Abstractions/NexusMods.Abstractions.Games.Diagnostics/Writer/DiagnosticWriterMode.cs:9]: I used to put `byte` on every enum out of habit writing binary serialization stuff, but in most cases it doesn't matter due to alignment.

**Comment by erri120** [comment]
> Quick Question:
> 
>     * Is the `Mode` intended to be dynamically mutated?
> 

Nope.

**Review by Sewer56 — APPROVED** [review]

**Pull request merged**